src/libxtp/matrixfreeoperator.cc

```diff
@@ -24,21 +24,59 @@ namespace xtp {
 
 Eigen::VectorXd MatrixFreeOperator::diagonal() const {
   Eigen::VectorXd D = Eigen::VectorXd::Zero(_size);
-#pragma omp parallel for
-  for (int i = 0; i < _size; i++) {
-    Eigen::RowVectorXd row_data = this->row(i);
-    D(i) = row_data(i);
+  if (useRow()) {
+#pragma omp parallel for schedule(guided)
+    for (int i = 0; i < _size; i++) {
+      Eigen::RowVectorXd row_data = this->OperatorRow(i);
+      D(i) = row_data(i);
+    }
   }
+
+  if (useBlock()) {
+    int blocksize = getBlocksize();
+    if (size() % blocksize != 0) {
+      throw std::runtime_error("blocksize is not a multiple of matrix size");
+    }
+    int blocks = size() / blocksize;
+
+// this is inefficient if blocks<num_ofthreads
+#pragma omp parallel for schedule(guided)
+    for (int i = 0; i < blocks; i++) {
+      Eigen::MatrixXd block = OperatorBlock(i, i);
+      D.segment(i * blocksize, blocksize) += block.diagonal();
```

Comment:
why += and not =

Reply:
I guess because you first extract the diag of the row operator and then add up the diag of the block operator .